### PR TITLE
Add migration document re differences between PER-CS v1.0 and PER-CS v2.0

### PR DESCRIPTION
**Pull request by kenguest**
When discussing differences between PSR-12 and PER-CS it is useful to have a summarised document detailing those changes.
This document is provided for that reason, and should be possible to use it to derive action lists for the likes of phpcs etc.

The scope of this document is to solely describe changes between the two aforementioned specs (PSR-12 and PER-CS v2.0), when required a new document would describe changes between PER-CS v2.0 and the next version - much in the same way that there are "Migration Documents" for PHP from one version to the next version. (We don't direct people to just look at the changelog files for PHP, instead migration documents are handwritten...)

It is intended that authors of tools such as PHP CodeSniffer and PHP CS Fixer would use this document(s) to drive their action lists so they can fully support each PER-CS iteration without having to investigate the minutiae.

### CODE REVIEW

**Review by Jean85 — COMMENTED** [review]
Is there a reason why we're not highlighting that PER-CS 1.0 is basically the same as PSR-12? That would streamline the process, and make the progression toward new PER-CS versions clearer in terms of changelog.

Also, we already have a rough changelog in the release page: https://github.com/php-fig/per-coding-style/releases/tag/2.0.0
Do we want to change it? Or to take that list in any way and put it here?

**Comment by kenguest** [comment]
I've added a link to that auto-generated changelog now.
I think for some people a coherent section-by-section discussion of the changes is going to be easier for them to follow instead of either looking at the various commit messages or even looking a a unified diff of those changes, which I personally find distracting.

The aim of this document is to lower the barrier for tool authors so they can create action lists of what they need to do to support PER-CS v2.0.

Consider it our equivalent of the migration documents such as https://www.php.net/manual/en/migration81.php, https://www.php.net/manual/en/migration83.php etc - we don't assume or direct people to just look at the changelogs for the new versions of PHP so I think it would be similarly polite that we provide the same for PER-CS.

**Comment by KorvinSzanto** [comment]
> I think for some people a coherent section-by-section discussion of the changes is going to be easier for them to follow instead of either looking at the various commit messages or even looking a a unified diff of those changes

We can make the content for the release whatever we want, it doesn't need to be various commit messages or a diff.

File: migration-to-PERCS-v2.0.md
Comment (Crell) [inline migration-to-PERCS-v2.0.md:1]: The file name doesn't need to specify that it's PERCS, but should specify that it's from 1.0 to 2.0.

The PHP docs use `migration81.php`, `migration82.php`, etc. as the naming convention.  So maybe just `migration-1.0-2.0.md`?  Then the next will be `migration-2.0-2.1.md` (or similar)?

File: migration-to-PERCS-v2.0.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+## Summary
+
+PER-CS is the next evolution of the PSR set of Coding Standards from the
+PHP-FIG (Framework Interoperability Group). It extends the Coding Standards
+laid out in PSR-12 to the newest functionality added to PHP such as the match
+keyword, enums, attributes and more.
+
+This document describes the changes and additions on a section by section
+basis between PER-CS v2.0 and PER-CS v1.0 which is a direct equivalent of 
+PSR12 with very minor changes.
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:12]:
```suggestion
basis between PER-CS v2.0 and PER-CS v1.0 (which is a direct equivalent of 
PSR12 with very minor changes).
```

The "which" needs to be its own clause.

File: migration-to-PERCS-v2.0.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+It is derived in part from [a GitHub generated changelog](https://github.com/php-fig/per-coding-style/compare/1.0.0...2.0.0)
+and focuses on the changes in a section-by-section basis as it's focus is to be more readable.
+
+The intent of this document is to provide a summary of these changes that can
+then be used to drive action-lists for toolset producers to support PER-CS v2.0.
+
+Any section not specified in this document does not have any PER-CS specific changes.
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:20]:
```suggestion
This document is non-normative.  The published [2.0 PER-CS]() specification is the canonical source for the PER-CS formatting expectations.
```

File: migration-to-PERCS-v2.0.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+## [Section 2.6](https://www.php-fig.org/per/coding-style/#26-trailing-commas)
+Numerous constructs now allow a sequence of values to have an optional trailing
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:23]:
```suggestion
## [Section 2.6](https://www.php-fig.org/per/coding-style/#26-trailing-commas)

Numerous constructs now allow a sequence of values to have an optional trailing
```

File: migration-to-PERCS-v2.0.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+Named arguments and method chaining
+
+When making a method or function call, there MUST NOT be a space between
+the method or function name and the opening parenthesis, there MUST NOT
+be a space after the opening parenthesis, and there MUST NOT be a space
+before the closing parenthesis.
+
+In the argument list, there MUST NOT be a space before each comma,
+and there MUST be one space after each comma.
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:65]: Are these actually new in 2.0?  I thought those were both very old, like PSR-2 old.

File: migration-to-PERCS-v2.0.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+line is indented once. When doing so, the first item in the list MUST be
+on the next line, and there MUST be only one argument per line. A single
+argument being split across multiple lines (as might be the case with a
+closure or array) does not constitute splitting the argument list itself.
+
+If using named arguments, there MUST NOT be a space between the argument
+name and colon, and there MUST be a single space between the colon and
+the argument value.
+
+
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:77]:
```suggestion
```

File: migration-to-PERCS-v2.0.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+## [Section 4.8](https://www.php-fig.org/per/coding-style/#48-function-callable-references)
+
+Function callable references - there must not be whitespace surrounding the '...' operator ()
+
+
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:82]:
```suggestion
```

File: migration-to-PERCS-v2.0.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+## [Section 5.2](https://www.php-fig.org/per/coding-style/#52-switch-case-match)
+The match keyword is now covered 
+
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:85]:
```suggestion
## [Section 5.2](https://www.php-fig.org/per/coding-style/#52-switch-case-match)

The match keyword is now covered.
```

File: migration-to-PERCS-v2.0.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+## [Section 7.1](https://www.php-fig.org/per/coding-style/#71-short-closures)
+
+A new subsection about Short Closures.
+
+As with standard closures, 'fn' must not be succeeded by a space.
+The '=>' symbol MUST be preceeded and succeeded by a space. (No leading or trailing spaces)
+The semicolon at the end MUST NOT be preceeded by a leading space.
+The expression portion MAY be split to a subsequent line. If so, the => MUST be 
+included on the second line, and MUST be indented once.
+
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:96]:
```suggestion
```

File: migration-to-PERCS-v2.0.md [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+## [Section 9](https://www.php-fig.org/per/coding-style/#9-enumerations)
+
+Enums - enums, or rather enumerations are a layer on top of classes and class
+constants that provide a way to define a very limited set of possible values
+for a custom type. Enums are new to PHP since 8.1 and so could not be covered
+by PSR12.
+
+PER-CS addresses aspects on how standardise on coding Enums in new section 9.
+As Enums do not support inheritance non-public methods MUST use the private
+keyword.
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:107]: Most of this exposition is not necessary.  Let's keep it short.  "Enums are now covered."  Just like `match` didn't bother to describe inline as it's a totally new section, Enums are a totally new section that we can just link straight to.  If we want to include anything else here, it should be just a bulleted list and/or a short example.

File: migration-to-PERCS-v2.0.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        public const Huge = self::Large;
+    }
+    ?>
+
+## [Section 10](https://www.php-fig.org/per/coding-style/#10-heredoc-and-nowdoc)
+
+This is a new section about HereDocs and NowDocs.
+
+NowDoc SHOULD be used whereever possible. Heredoc MAY be used where a
+nowdoc does not satisfy.
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:148]:
```suggestion
nowdoc is not sufficient.
```

File: migration-to-PERCS-v2.0.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+the declaration is being used in. Subsequent lines in the heredoc
+or nowdoc MUST be indented once past the scope indentation they are
+declared in.
+
+The heredoc MUST be declared on the same line as the variable declaration it's being set against.
+The heredoc MUST be indented once past the indentation of the scope it's declared in.
+
+## [Section 11](https://www.php-fig.org/per/coding-style/#11-arrays)
+
+This too is a new section. It is about arrays.
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:160]:
```suggestion
This is a new section.
```

File: migration-to-PERCS-v2.0.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+Any section not specified in this document does not have any PER-CS specific changes.
+
+## [Section 2.6](https://www.php-fig.org/per/coding-style/#26-trailing-commas)
+Numerous constructs now allow a sequence of values to have an optional trailing
+comma:
+ * If the final item is on the same line then there MUST NOT be a trailing comma
+ * If the final item is not on the same line then there MUST be a trailing comma
+
+## [Section 4.4](https://www.php-fig.org/per/coding-style/#44-methods-and-functions)
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:28]: The heading titles should include what the section is, not just the section reference.  As long as there's a link, the section reference isn't all that useful.  A proper title is.

File: migration-to-PERCS-v2.0.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+There MUST NOT be more than one value assignment per line. 
+
+Value assignments MAY use a single line or multiple lines.
+
+## [Section 12](https://www.php-fig.org/per/coding-style/#12-attributes)
+
+This is a new section about attributes. Attributes are new in PHP 8 and
+offer the ability to add structured, machine-readable metadata information on 
+declarations in code (https://www.php.net/manual/en/language.attributes.overview.php)
```
Comment (Crell) [inline migration-to-PERCS-v2.0.md:182]: Again, there's no point in calling out that they're new.  I'm undecided about manual links.

Comment (Crell) [reply migration-to-PERCS-v2.0.md:182]: This whole section feels like it's just duplicating the actual spec document.  That is prone to confusion if the wording is slightly different in each case.

File: migration-to-PERCS-v2.0.md
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:1]: If we follow the PHP docs naming scheme for migration docs, then maybe we leave out the version being migrated from and just call it migration2.0.md?

File: migration-to-PERCS-v2.0.md (hunk 1 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:12]: done

File: migration-to-PERCS-v2.0.md (hunk 2 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:20]: done

File: migration-to-PERCS-v2.0.md (hunk 3 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:23]: done

File: migration-to-PERCS-v2.0.md (hunk 5 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:77]: done

File: migration-to-PERCS-v2.0.md (hunk 6 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:82]: done

File: migration-to-PERCS-v2.0.md (hunk 7 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:85]: done

File: migration-to-PERCS-v2.0.md (hunk 9 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:107]: Agreed

File: migration-to-PERCS-v2.0.md (hunk 10 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:148]: done

File: migration-to-PERCS-v2.0.md (hunk 11 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:160]: done

File: migration-to-PERCS-v2.0.md (hunk 13 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:182]: I reduced the introductory paragraph for this section, removing the exposition and the link to the manual.
People can do an internet search...

File: migration-to-PERCS-v2.0.md (hunk 12 above)
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:28]: Expanded the heading titles for clarity.

File: migration-to-PERCS-v2.0.md
Comment (kenguest) [reply migration-to-PERCS-v2.0.md:1]: Thought on this some more and renamed the file to migration-2.0.md

**Comment by kenguest** [comment]
@Crell  - I've addressed all issues you raised now, if you'd like to give it another look over please?

File: migration-2.0.md [hunk 14]
```diff
@@ -0,0 +1,196 @@
+# Migrating from PER-CS v1.0 (PSR-12) to PER-CS v2.0 ###
+
+## Summary
+
+PER-CS is the next evolution of the PSR set of Coding Standards from the
+PHP-FIG (Framework Interoperability Group). It extends the Coding Standards
+laid out in PSR-12 to the newest functionality added to PHP such as the match
+keyword, enums, attributes and more.
```
Comment (samdark) [inline migration-2.0.md:8]:
```suggestion
keyword, enums, attributes, and more.
```

File: migration-2.0.md [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+
+This document describes the changes and additions on a section by section
+basis between PER-CS v2.0 and PER-CS v1.0 (which is a direct equivalent of 
+PSR12 with very minor changes).
+
+It is derived in part from [a GitHub generated changelog](https://github.com/php-fig/per-coding-style/compare/1.0.0...2.0.0)
```
Comment (samdark) [inline migration-2.0.md:14]:
```suggestion
It is derived in part from [a GitHub-generated changelog](https://github.com/php-fig/per-coding-style/compare/1.0.0...2.0.0)
```

File: migration-2.0.md [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+and focuses on the changes in a section-by-section basis as it's focus is to be more readable.
+
+The intent of this document is to provide a summary of these changes that can
```
Comment (samdark) [inline migration-2.0.md:17]:
```suggestion
This document intends to provide a summary of these changes that can
```

File: migration-2.0.md [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+
+If an attribute has no arguments, the () MUST be omitted.
+
+The closing attribute block indicator ] MUST follow the last character
+of the attribute name or the closing ) of its argument list, with no
+preceding space.
+
+The construct #[...] is referred to as an "attribute block" in this document.
+
+Attributes on classes, methods, functions, constants and properties
```
Comment (samdark) [inline migration-2.0.md:168]:
```suggestion
Attributes on classes, methods, functions, constants, and properties
```

File: migration-2.0.md [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+MUST be placed on their own line, immediately prior to the structure
```
Comment (samdark) [inline migration-2.0.md:169]:
```suggestion
MUST be placed on their own line, immediately before the structure
```

File: migration-2.0.md [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+being described.
+
+For attributes on parameters, if the parameter list is presented on a
+single line, the attribute MUST be placed inline with the parameter it
+describes, separated by a single space. If the parameter list is split
+into multiple lines for any reason, the attribute MUST be placed on its
+own line prior to the parameter, indented the same as the parameter. If
```
Comment (samdark) [inline migration-2.0.md:176]:
```suggestion
own line before the parameter, indented the same as the parameter. If
```

File: migration-2.0.md [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+the parameter list is split into multiple lines, a blank line MAY be
+included between one parameter and the attributes of the following
+parameter in order to aid readability.
```
Comment (samdark) [inline migration-2.0.md:179]:
```suggestion
parameter to aid readability.
```

**Comment by Crell** [comment]
I am still torn on whether we should include as much detail for the new sections.  What's the value of providing more text to read beyond "there's a new section on arrays.  Go read it."  Literally every word in a new section is something an implementer needs to care about, so the summary is essentially redundant with the spec.

**Comment by samdark** [comment]
@Crell could be turned into a blog post if we don't want it in meta-doc. But I think such meta-doc is useful to increase adoption.

**Comment by kenguest** [comment]
@samdark I've tightened things up as you suggested.

File: migration-2.0.md [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+## [Section 5.2 - Switch, Case, Match](https://www.php-fig.org/per/coding-style/#52-switch-case-match)
+
+The match keyword is now covered.
+
+## [Section 7.1 - Short Closures](https://www.php-fig.org/per/coding-style/#71-short-closures)
+
+A new subsection about Short Closures.
+
+As with standard closures, 'fn' must not be succeeded by a space.
+The '=>' symbol MUST be preceeded and succeeded by a space. (No leading or trailing spaces)
```
Comment (KorvinSzanto) [inline migration-2.0.md:77]:
```suggestion
The '=>' symbol MUST be preceded and succeeded by a space (No leading or trailing spaces).
```

File: migration-2.0.md [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+PER-CS is the next evolution of the PSR set of Coding Standards from the
+PHP-FIG (Framework Interoperability Group). It extends the Coding Standards
+laid out in PSR-12 to the newest functionality added to PHP such as the match
+keyword, enums, attributes, and more.
+
+This document describes the changes and additions on a section by section
+basis between PER-CS v2.0 and PER-CS v1.0 (which is a direct equivalent of 
+PSR12 with very minor changes).
+
+It is derived in part from [a GitHub-generated changelog](https://github.com/php-fig/per-coding-style/compare/1.0.0...2.0.0)
```
Comment (KorvinSzanto) [inline migration-2.0.md:14]:
```suggestion
It is derived in part from [a GitHub-generated diff](https://github.com/php-fig/per-coding-style/compare/1.0.0...2.0.0#files_bucket)
```

**Comment by kenguest** [comment]
Updated as requested by @Crell  earlier today...

**Comment by KorvinSzanto** [comment]
Can you update the code blocks to be 3 backticks rather than indented by 4 spaces and remove the closing tags? 

```
```php
```
enables PHP syntax highlighting work. See examples from the spec: https://github.com/php-fig/per-coding-style/blob/master/spec.md?plain=1#L266-L275

File: migration-2.0.md [hunk 23]
```diff
@@ -0,0 +1,233 @@
...
+
+Method chaining MAY be put on separate lines, where each subsequent line is indented once. When doing so, the first method MUST be on the next line.
+
+## [Section 4.8 - Function Callable References](https://www.php-fig.org/per/coding-style/#48-function-callable-references)
+
+Function callable references - there must not be whitespace surrounding the '...' operator ()
+
+## [Section 5.2 - Switch, Case, Match](https://www.php-fig.org/per/coding-style/#52-switch-case-match)
+
+The match keyword is now covered.
```
Comment (Crell) [inline migration-2.0.md:71]: Example of match, for consistency?

File: migration-2.0.md [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+
+Furthermore, all keywords must be on a single line and MUST be separated
+by a single space.
+
+## [Section 4.7 - Method and Function Calls](https://www.php-fig.org/per/coding-style/#47-method-and-function-calls)
+
+If using named arguments, there MUST NOT be a space between the argument name and the colon, 
+and there MUST be a single space between the colon and the argument value.
+
+Method chaining MAY be put on separate lines, where each subsequent line is indented once. When doing so, the first method MUST be on the next line.
```
Comment (Crell) [inline migration-2.0.md:63]: Example(s) for this section?

**Review by KorvinSzanto — APPROVED** [review]
LGTM :+1:

**Review by Crell — APPROVED** [review]

**Pull request merged**